File: apkEditor/src/main/java/com/mcal/apkeditor/smali/AsyncDecodeTask.java
package com.mcal.apkeditor.smali;

import android.annotation.SuppressLint;
import android.content.Context;
import android.os.AsyncTask;
import android.util.Log;

import androidx.annotation.NonNull;

import com.mcal.apkeditor.pro.DexDecoder;
import com.mcal.common.utils.IOUtils;
import com.mcal.common.utils.SDCard;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.Enumeration;
import java.util.List;
import java.util.zip.ZipEntry;
import java.util.zip.ZipFile;

public class AsyncDecodeTask extends AsyncTask<Void, Integer, Boolean> {

    @SuppressLint("StaticFieldLeak")
    private final Context context;
    private final IDecodeTaskCallback callback;
    private final String decodeRootPath;
    private final String apkPath;
    // Record all dex file paths which are extracted from apk file
    private final List<String> dexFileList = new ArrayList<>();
    private String strError;
    private String strWarning;

    public AsyncDecodeTask(Context context, String apkPath,
                           String decodeRootPath, IDecodeTaskCallback callback) {
        this.context = context;
        this.callback = callback;
        this.apkPath = apkPath;
        this.decodeRootPath = decodeRootPath;
    }

    @Override
    protected void onPreExecute() {
        if (callback != null) {
            callback.dexDecodingStarted();
        }
    }

    private void prepareDexFiles() throws Exception {
        String tmpDirectory = SDCard.makeDir(context, "tmp");

        ZipFile zipFile = null;
        try {
            zipFile = new ZipFile(apkPath);
            Enumeration<? extends ZipEntry> entries = zipFile.entries();
            while (entries.hasMoreElements()) {
                ZipEntry entry = entries.nextElement();
                String name = entry.getName();
                if (name.endsWith(".dex") && !name.contains("/")) {
                    unzipDex2File(zipFile, entry, tmpDirectory + name);
                }
            }
        } catch (IOException e1) {
            e1.printStackTrace();
        } finally {
            if (zipFile != null) {
                try {
                    zipFile.close();
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
        }
    }

    private void prepareMainDex() throws Exception {
        String tmpDirectory = SDCard.makeDir(context, "tmp");

        ZipFile zipFile = null;
        try {
            String name = "classes.dex";
            zipFile = new ZipFile(apkPath);
            ZipEntry entry = zipFile.getEntry(name);
            unzipDex2File(zipFile, entry, tmpDirectory + name);
        } catch (IOException e1) {
            e1.printStackTrace();
        } finally {
            if (zipFile != null) {
                try {
                    zipFile.close();
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
        }
    }

    private void unzipDex2File(@NonNull ZipFile zipFile, ZipEntry entry, String filePath)
            throws IOException {
        InputStream in = null;
        FileOutputStream out = null;
        try {
            in = zipFile.getInputStream(entry);
            out = new FileOutputStream(filePath);
            IOUtils.copy(in, out);
            this.dexFileList.add(filePath);
        } finally {
            if (in != null) {
                try {
                    in.close();
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
            if (out != null) {
                try {
                    out.close();
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
        }
    }

    @Override
    protected Boolean doInBackground(Void... params) {
        try {
            doAllJobs();
            return true;
        } catch (Exception e) {
            this.strError = e.getMessage();
            return false;
        }
    }

    public void doAllJobs() throws Exception {
        prepareDexFiles();
        decodeDexFiles();
        removeDexFiles();
    }

    // Only decode the classes.dex
    public void decodeMainDex() throws Exception {
        prepareMainDex();
        decodeDexFiles();
        removeDexFiles();
    }

    private void removeDexFiles() {
        for (String filePath : this.dexFileList) {
            File f = new File(filePath);
            f.delete();
        }
    }

    private void decodeDexFiles() {
        for (String dexFilePath : this.dexFileList) {
            DexDecoder decoder = new DexDecoder(dexFilePath);

            String directory = this.decodeRootPath + "/smali";
            // Not the default dex file
            if (!dexFilePath.endsWith("/classes.dex")) {
                int position = dexFilePath.lastIndexOf("/");
                String dexName = dexFilePath.substring(position + 1,
                        dexFilePath.length() - 4);
                directory = this.decodeRootPath + "/smali_" + dexName;
            }
            createDirectoryIfNotExist(directory);

            try {
                decoder.dex2smali(directory);
            } catch (Exception e) {
                e.printStackTrace();
            }


            if (this.strWarning == null) {
                this.strWarning = decoder.getWarning();
            }
        }
    }

    private void createDirectoryIfNotExist(String directory) {
        File dir = new File(directory);
        if (dir.exists()) {
            dir.mkdir();
        }
    }

    @Override
    protected void onPostExecute(Boolean result) {
        if (callback != null) {
            if (result) {
                for (File file : new File(decodeRootPath).listFiles()) {
                    if (file.getName().endsWith(".dex")) {
                       if(file.delete()) {
                           Log.e(getClass().getName(), file + " deleted");
                       }
                    }
                }
                callback.dexDecodingFinished(true, null, strWarning);
            } else {
                callback.dexDecodingFinished(false, strError, null);
            }
        }
    }

    public interface IDecodeTaskCallback {
        void dexDecodingStarted();

        void dexDecodingFinished(boolean result, String strError,
                                        String strWarning);
    }
}

File: apkEditor/src/main/java/com/mcal/apkeditor/smali/AsyncDecodeTask.kt
package com.mcal.apkeditor.smali

import android.content.Context
import android.util.Log
import com.mcal.apkeditor.pro.DexDecoder
import com.mcal.common.utilsOld.IOUtils
import com.mcal.common.utilsOld.SDCard
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.io.InputStream
import java.util.zip.ZipEntry
import java.util.zip.ZipFile
import kotlin.coroutines.CoroutineContext

class AsyncDecodeTask(
    private val context: Context,
    private val mApkPath: String,
    private val mDecodeRootPath: String,
    private val mCallback: IDecodeTaskCallback?
) : CoroutineScope {
    // Record all dex file paths which are extracted from apk file
    private val dexFileList: MutableList<String> = ArrayList()
    private var strError: String? = null
    private var strWarning: String? = null

    override val coroutineContext: CoroutineContext
        get() = Dispatchers.Main

    fun execute() = launch {
        onPreExecute()
        val result = doInBackground()
        onPostExecute(result)
    }

    private fun onPreExecute() {
        mCallback?.dexDecodingStarted()
    }

    private suspend fun doInBackground(): Boolean = withContext(Dispatchers.IO) {
        return@withContext try {
            CoroutineScope(Dispatchers.IO).launch {

            }
            doAllJobs()
            true
        } catch (e: Exception) {
            strError = e.message
            false
        }
    }

    private fun onPostExecute(result: Boolean) {
        mCallback?.let { callback ->
            if (result) {
                File(mDecodeRootPath).listFiles()?.let { listFile ->
                    for (file in listFile) {
                        if (file.name.endsWith(".dex")) {
                            if (file.delete()) {
                                Log.e(javaClass.name, "$file deleted")
                            }
                        }
                    }
                }
                callback.dexDecodingFinished(true, null, strWarning)
            } else {
                callback.dexDecodingFinished(false, strError, null)
            }
        }
    }

    @Throws(Exception::class)
    private fun prepareDexFiles() {
        val tmpDirectory = SDCard.makeDir(context, "tmp")
        var zipFile: ZipFile? = null
        try {
            zipFile = ZipFile(mApkPath)
            val entries = zipFile.entries()
            while (entries.hasMoreElements()) {
                val entry = entries.nextElement()
                val name = entry.name
                if (name.endsWith(".dex") && !name.contains("/")) {
                    unzipDex2File(zipFile, entry, tmpDirectory + name)
                }
            }
        } catch (e1: IOException) {
            e1.printStackTrace()
        } finally {
            zipFile?.let {
                try {
                    zipFile.close()
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }
        }
    }

    @Throws(Exception::class)
    private fun prepareMainDex() {
        val tmpDirectory = SDCard.makeDir(context, "tmp")
        var zipFile: ZipFile? = null
        try {
            val name = "classes.dex"
            zipFile = ZipFile(mApkPath)
            val entry = zipFile.getEntry(name)
            unzipDex2File(zipFile, entry, tmpDirectory + name)
        } catch (e1: IOException) {
            e1.printStackTrace()
        } finally {
            zipFile?.let {
                try {
                    zipFile.close()
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }
        }
    }

    @Throws(IOException::class)
    private fun unzipDex2File(zipFile: ZipFile, entry: ZipEntry, filePath: String) {
        var input: InputStream? = null
        var output: FileOutputStream? = null
        try {
            input = zipFile.getInputStream(entry)
            output = FileOutputStream(filePath)
            IOUtils.copy(input, output)
            dexFileList.add(filePath)
        } finally {
            input?.let {
                try {
                    input.close()
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }
            output?.let {
                try {
                    output.close()
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }
        }
    }

    @Throws(Exception::class)
    fun doAllJobs() {
        prepareDexFiles()
        decodeDexFiles()
        removeDexFiles()
    }

    // Only decode the classes.dex
    @Throws(Exception::class)
    fun decodeMainDex() {
        prepareMainDex()
        decodeDexFiles()
        removeDexFiles()
    }

    private fun removeDexFiles() {
        for (filePath in dexFileList) {
            val f = File(filePath)
            f.delete()
        }
    }

    private fun decodeDexFiles() {
        for (dexFilePath in dexFileList) {
            val decoder = DexDecoder(dexFilePath)
            var directory = "$mDecodeRootPath/smali"
            // Not the default dex file
            if (!dexFilePath.endsWith("/classes.dex")) {
                val position = dexFilePath.lastIndexOf("/")
                val dexName = dexFilePath.substring(
                    position + 1,
                    dexFilePath.length - 4
                )
                directory = "$mDecodeRootPath/smali_$dexName"
            }
            createDirectoryIfNotExist(directory)
            try {
                decoder.dex2smali(directory)
            } catch (e: Exception) {
                e.printStackTrace()
            }
            if (strWarning == null) {
                strWarning = decoder.warning
            }
        }
    }

    private fun createDirectoryIfNotExist(directory: String) {
        val dir = File(directory)
        if (dir.exists()) {
            dir.mkdir()
        }
    }

    interface IDecodeTaskCallback {
        fun dexDecodingStarted()
        fun dexDecodingFinished(
            result: Boolean, strError: String?,
            strWarning: String?
        )
    }
}